finality-aleph/src/party/mod.rs

```diff
@@ -209,9 +208,12 @@ where
 
     debug!(target: "afa", "Consensus network has started.");
 
+    let last_finalized_number = client.info().finalized_number;
+    let starting_session =
+        session_id_from_block_num::<B>(last_finalized_number, session_period);
     let authorities = client
         .runtime_api()
-        .authorities(&BlockId::Number(<NumberFor<B>>::saturated_from(0u32)))
+        .authorities(&BlockId::Number(<NumberFor<B>>::saturated_from(starting_session.0)))
         .expect("We should know authorities for the first session");
```

Comment:
I would much rather like this and the starting session to be computed inside `run()` and not here. I know that you have a field `.authorities` in `ConsensusParty` that tracks authorities and that's why you must do that. I find this field `.authorities` super confusing in the current version, as it really means `.next_session_authorities`, but even that not quite as it depends on the timing. So either remove this field completely (I don't know if it's doable) or replace this field by something else. For instance there is `session_authorities` already that holds authorities by `session_id`, if you cannot reuse it, just create a separate map for that. Having a map, instead of a single field is much better because you can never make a mistake in interpreting what "current authorities" or "next authorities" mean, since the numbering is absolute and not relative to the session in progress.

If you end up making a map, you should also remember to prune it, see `prune_session_data`.

Reply:
And yes, even if the map must hold max 2 elements at a time, it's not a big downside compared to the clarity it adds.

Reply:
Ok, we had discussion exactly about it with @timorl few days ago XD